crates/codegen/schema/src/types/precedence_parser.rs

```diff
@@ -0,0 +1,51 @@
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+use super::parser::ParserDefinition;
+
+#[allow(dead_code)]
+pub type PrecedenceParserRef = std::rc::Rc<PrecedenceParser>;
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields)]
+pub struct PrecedenceParser {
+    #[serde(default)]
+    pub name: Option<String>,
+    #[serde(flatten)]
+    pub definition: PrecedenceParserDefinition,
+}
+
+#[derive(Clone, Debug, Deserialize, Serialize, JsonSchema, PartialEq, Eq, Hash)]
+#[serde(deny_unknown_fields, rename_all = "camelCase")]
+pub struct PrecedenceParserDefinition {
+    #[schemars(title = "Operator Definitions")]
+    pub operators: Vec<OperatorDefinition>,
+
+    #[schemars(title = "Primary Expression References")]
```

Comment:
Is `primary_expressions` array intended to be alternatives? What is the use case?

I see the grammar still defines `PrimaryExpression` as a single `reference:` to a `choice:`, instead of inlining them here, which I think is better, since it doesn't duplicate its contents between multiple versions.